src/lm/arpa-file-parser.h

```diff
@@ -86,9 +91,35 @@ class ArpaFileParser {
   void Read(std::istream &is, bool binary);
 
   /// Parser options.
-  const ArpaParseOptions& Options() const { return options_; }
+  const ArpaParseOptions &Options() const { return options_; }
 
  protected:
+  /// A status value returned from ConsumeNGram() overload.
+  class Status {
+   public:
+    /// Return the warninig status to print a warninig and continue.
+    static Status Warning(const std::string& message) {
+      return Status(kWarninig, message); }
```

Comment:
The global variable is specific to this of ARPA-reading code, which is only
accessed from this one program.  Or having it in the options class or in
the class itself would be fine with me- whatever you prefer.

If you really need the line number, you can make it an additional argument
to the function where it's needed.

That way all the errors and warnings can be printed exactly where they
originate.  I'm not going to be persuaded easily on this issue.


On Sun, Apr 3, 2016 at 11:59 PM, Kirill Katsnelson <notifications@github.com
> wrote:

> In src/lm/arpa-file-parser.h
> <https://github.com/kaldi-asr/kaldi/pull/649#discussion_r58326624>:
>
> >
> >   protected:
> > +  /// A status value returned from ConsumeNGram() overload.
> > +  class Status {
> > +   public:
> > +    /// Return the warninig status to print a warninig and continue.
> > +    static Status Warning(const std::string& message) {
> > +      return Status(kWarninig, message); }
>
> Well, I did not do that for the love of classes. The diagnostic was really
> poor: I could print a line number in ARPA file, but if that comes out of
> pipeline, good luck finding it. So I need to communicate back from
> ConsumeNGram what it decided to do with it: accept, accept with a warning
> or die fatally. Formatting is done in the common class, and is actually
> useful:for locating the problem:
>
> line 21 [-1.5       a β -3.2] skipped: word 'β' not in symbol table.
> line 24 [-0.03333        <s> b b b] skipped: no parent (n-1)-gram exists
>
> Rate limiting *by itself* is not a problem, the counter is a private
> member of the ARPA parser class. The question is *what* to count. I would
> really hate to increment a global variable defined in one file from two
> others every time I print KALDI_WARN. It's probably even worse.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/649/files/939259c797b78d91abfcfe582813d8cd3afccbf8#r58326624>
>